File: packages/prover-sdk/src/circuits/registry.ts
/**
 * @bitsage/prover-sdk - Circuit Registry
 *
 * Registry of all supported circuits with their configurations.
 * Maps circuit types to their metadata, verifier addresses, and timing estimates.
 */

import { CircuitType, CircuitConfig, ProofMode } from '../types';

// ============================================================================
// Verifier Contract Addresses (Sepolia)
// ============================================================================

const VERIFIERS = {
  STWO_VERIFIER: '0x52963fe2f1d2d2545cbe18b8230b739c8861ae726dc7b6f0202cc17a369bd7d',
  PRIVACY_POOLS: '0xd85ad03dcd91a075bef0f4226149cb7e43da795d2c1d33e3227c68bfbb78a7',
  PRIVACY_ROUTER: '0x7d1a6c242a4f0573696e117790f431fd60518a000b85fe5ee507456049ffc53',
  CONFIDENTIAL_SWAP: '0x29516b3abfbc56fdf0c1f136c971602325cbabf07ad8f984da582e2106ad2af',
} as const;

// ============================================================================
// Circuit Configurations
// ============================================================================

const CIRCUIT_CONFIGS: Record<CircuitType, CircuitConfig> = {
  // === Compute Proofs (GPU Workers) ===
  [CircuitType.AI_INFERENCE]: {
    id: CircuitType.AI_INFERENCE,
    name: 'AI Inference',
    defaultMode: ProofMode.WORKER_GPU,
    verifierAddress: VERIFIERS.STWO_VERIFIER,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 45000,
      [ProofMode.WORKER_GPU]: 5000,
    },
    provingKeySize: 52_428_800, // 50MB
  },

  [CircuitType.DATA_PIPELINE]: {
    id: CircuitType.DATA_PIPELINE,
    name: 'Data Pipeline',
    defaultMode: ProofMode.WORKER_GPU,
    verifierAddress: VERIFIERS.STWO_VERIFIER,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 35000,
      [ProofMode.WORKER_GPU]: 4000,
    },
    provingKeySize: 41_943_040, // 40MB
  },

  [CircuitType.ML_TRAINING]: {
    id: CircuitType.ML_TRAINING,
    name: 'ML Training',
    defaultMode: ProofMode.WORKER_GPU,
    verifierAddress: VERIFIERS.STWO_VERIFIER,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 120000,
      [ProofMode.WORKER_GPU]: 15000,
    },
    provingKeySize: 104_857_600, // 100MB
  },

  [CircuitType.GENERIC_COMPUTE]: {
    id: CircuitType.GENERIC_COMPUTE,
    name: 'Generic Compute',
    defaultMode: ProofMode.WORKER_GPU,
    verifierAddress: VERIFIERS.STWO_VERIFIER,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 25000,
      [ProofMode.WORKER_GPU]: 3000,
    },
    provingKeySize: 31_457_280, // 30MB
  },

  // === Privacy Proofs (TEE Enclaves) ===
  [CircuitType.PRIVACY_WITHDRAW]: {
    id: CircuitType.PRIVACY_WITHDRAW,
    name: 'Privacy Withdraw',
    defaultMode: ProofMode.TEE_ASSISTED,
    verifierAddress: VERIFIERS.PRIVACY_POOLS,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 12000,
      [ProofMode.TEE_ASSISTED]: 3000,
    },
    provingKeySize: 20_971_520, // 20MB
  },

  [CircuitType.PRIVACY_TRANSFER]: {
    id: CircuitType.PRIVACY_TRANSFER,
    name: 'Privacy Transfer',
    defaultMode: ProofMode.TEE_ASSISTED,
    verifierAddress: VERIFIERS.PRIVACY_ROUTER,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 15000,
      [ProofMode.TEE_ASSISTED]: 4000,
    },
    provingKeySize: 26_214_400, // 25MB
  },

  [CircuitType.CONFIDENTIAL_SWAP]: {
    id: CircuitType.CONFIDENTIAL_SWAP,
    name: 'Confidential Swap',
    defaultMode: ProofMode.TEE_ASSISTED,
    verifierAddress: VERIFIERS.CONFIDENTIAL_SWAP,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 18000,
      [ProofMode.TEE_ASSISTED]: 5000,
    },
    provingKeySize: 31_457_280, // 30MB
  },

  // === Lightweight Proofs (Client WASM) ===
  [CircuitType.MERKLE_MEMBERSHIP]: {
    id: CircuitType.MERKLE_MEMBERSHIP,
    name: 'Merkle Membership',
    defaultMode: ProofMode.CLIENT_WASM,
    verifierAddress: VERIFIERS.PRIVACY_POOLS,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 5000,
    },
    provingKeySize: 10_485_760, // 10MB
  },

  [CircuitType.RANGE_PROOF]: {
    id: CircuitType.RANGE_PROOF,
    name: 'Range Proof',
    defaultMode: ProofMode.CLIENT_WASM,
    verifierAddress: VERIFIERS.PRIVACY_ROUTER,
    estimatedTimeMs: {
      [ProofMode.CLIENT_WASM]: 3000,
    },
    provingKeySize: 5_242_880, // 5MB
  },
};

// ============================================================================
// Circuit Registry API
// ============================================================================

export class CircuitRegistry {
  private static configs = new Map<CircuitType, CircuitConfig>(
    Object.entries(CIRCUIT_CONFIGS).map(([, config]) => [config.id, config])
  );

  /**
   * Get circuit configuration by type
   */
  static get(circuit: CircuitType): CircuitConfig | undefined {
    return this.configs.get(circuit);
  }

  /**
   * Get all circuit configurations
   */
  static getAll(): CircuitConfig[] {
    return Array.from(this.configs.values());
  }

  /**
   * Get circuits by mode
   */
  static getByMode(mode: ProofMode): CircuitConfig[] {
    return this.getAll().filter(c => c.defaultMode === mode);
  }

  /**
   * Get circuit by name (case-insensitive)
   */
  static getByName(name: string): CircuitConfig | undefined {
    return this.getAll().find(c => c.name.toLowerCase() === name.toLowerCase());
  }

  /**
   * Get total proving key size for a set of circuits
   */
  static getTotalProvingKeySize(circuits: CircuitType[]): number {
    return circuits.reduce((sum, c) => {
      const config = this.get(c);
      return sum + (config?.provingKeySize ?? 0);
    }, 0);
  }

  /**
   * Get estimated proof time for circuit + mode
   */
  static getEstimatedTime(circuit: CircuitType, mode: ProofMode): number | undefined {
    const config = this.get(circuit);
    return config?.estimatedTimeMs[mode];
  }

  /**
   * Check if circuit supports a given mode
   */
  static supportsMode(circuit: CircuitType, mode: ProofMode): boolean {
    const config = this.get(circuit);
    if (!config) return false;

    return config.estimatedTimeMs[mode] !== undefined;
  }

  /**
   * Get CDN path for proving key
   */
  static getProvingKeyPath(circuit: CircuitType, cdnBase: string): string {
    const config = this.get(circuit);
    if (!config) {
      throw new Error(`Unknown circuit: ${circuit}`);
    }

    const circuitName = config.name.toLowerCase().replace(/\s+/g, '-');
    return `${cdnBase}/${circuitName}/proving-key.bin`;
  }

  /**
   * Register a custom circuit (for extensions)
   */
  static register(config: CircuitConfig): void {
    this.configs.set(config.id, config);
  }
}

// ============================================================================
// Circuit Type Guards
// ============================================================================

/**
 * Check if circuit requires privacy (TEE) mode
 */
export function isPrivacyCircuit(circuit: CircuitType): boolean {
  return [
    CircuitType.PRIVACY_WITHDRAW,
    CircuitType.PRIVACY_TRANSFER,
    CircuitType.CONFIDENTIAL_SWAP,
  ].includes(circuit);
}

/**
 * Check if circuit is compute-intensive (GPU preferred)
 */
export function isComputeCircuit(circuit: CircuitType): boolean {
  return [
    CircuitType.AI_INFERENCE,
    CircuitType.DATA_PIPELINE,
    CircuitType.ML_TRAINING,
    CircuitType.GENERIC_COMPUTE,
  ].includes(circuit);
}

/**
 * Check if circuit is lightweight (WASM suitable)
 */
export function isLightweightCircuit(circuit: CircuitType): boolean {
  return [
    CircuitType.MERKLE_MEMBERSHIP,
    CircuitType.RANGE_PROOF,
  ].includes(circuit);
}

File: rust-node/src/prover/registry.rs
//! Circuit Registry
//!
//! Registry of supported circuits and their configurations.

use std::collections::HashMap;

use super::types::*;
use super::router::CircuitConfig;
use crate::contracts::contracts;

/// Circuit registry
pub struct CircuitRegistry {
    circuits: HashMap<CircuitType, CircuitConfig>,
}

impl CircuitRegistry {
    pub fn new() -> Self {
        let mut circuits = HashMap::new();

        // Get contract addresses from centralized config
        let addrs = contracts();

        // Compute proofs
        circuits.insert(CircuitType::AiInference, CircuitConfig {
            id: CircuitType::AiInference,
            name: "AI Inference".to_string(),
            default_mode: ProofMode::WorkerGpu,
            verifier_address: addrs.stwo_verifier.clone(),
            estimated_time_ms: 5000,
            proving_key_size: 52_428_800, // 50MB
        });

        circuits.insert(CircuitType::DataPipeline, CircuitConfig {
            id: CircuitType::DataPipeline,
            name: "Data Pipeline".to_string(),
            default_mode: ProofMode::WorkerGpu,
            verifier_address: addrs.stwo_verifier.clone(),
            estimated_time_ms: 4000,
            proving_key_size: 41_943_040, // 40MB
        });

        circuits.insert(CircuitType::MlTraining, CircuitConfig {
            id: CircuitType::MlTraining,
            name: "ML Training".to_string(),
            default_mode: ProofMode::WorkerGpu,
            verifier_address: addrs.stwo_verifier.clone(),
            estimated_time_ms: 15000,
            proving_key_size: 104_857_600, // 100MB
        });

        circuits.insert(CircuitType::GenericCompute, CircuitConfig {
            id: CircuitType::GenericCompute,
            name: "Generic Compute".to_string(),
            default_mode: ProofMode::WorkerGpu,
            verifier_address: addrs.stwo_verifier.clone(),
            estimated_time_ms: 3000,
            proving_key_size: 31_457_280, // 30MB
        });

        // Privacy proofs
        circuits.insert(CircuitType::PrivacyWithdraw, CircuitConfig {
            id: CircuitType::PrivacyWithdraw,
            name: "Privacy Withdraw".to_string(),
            default_mode: ProofMode::TeeAssisted,
            verifier_address: addrs.privacy_pools.clone(),
            estimated_time_ms: 3000,
            proving_key_size: 20_971_520, // 20MB
        });

        circuits.insert(CircuitType::PrivacyTransfer, CircuitConfig {
            id: CircuitType::PrivacyTransfer,
            name: "Privacy Transfer".to_string(),
            default_mode: ProofMode::TeeAssisted,
            verifier_address: addrs.privacy_router.clone(),
            estimated_time_ms: 4000,
            proving_key_size: 26_214_400, // 25MB
        });

        circuits.insert(CircuitType::ConfidentialSwap, CircuitConfig {
            id: CircuitType::ConfidentialSwap,
            name: "Confidential Swap".to_string(),
            default_mode: ProofMode::TeeAssisted,
            verifier_address: addrs.confidential_swap.clone(),
            estimated_time_ms: 5000,
            proving_key_size: 31_457_280, // 30MB
        });

        // Lightweight proofs
        circuits.insert(CircuitType::MerkleMembership, CircuitConfig {
            id: CircuitType::MerkleMembership,
            name: "Merkle Membership".to_string(),
            default_mode: ProofMode::ClientWasm,
            verifier_address: addrs.privacy_pools.clone(),
            estimated_time_ms: 5000,
            proving_key_size: 10_485_760, // 10MB
        });

        circuits.insert(CircuitType::RangeProof, CircuitConfig {
            id: CircuitType::RangeProof,
            name: "Range Proof".to_string(),
            default_mode: ProofMode::ClientWasm,
            verifier_address: addrs.privacy_router.clone(),
            estimated_time_ms: 3000,
            proving_key_size: 5_242_880, // 5MB
        });

        Self { circuits }
    }

    /// Get circuit configuration
    pub fn get(&self, circuit: CircuitType) -> Option<CircuitConfig> {
        self.circuits.get(&circuit).cloned()
    }

    /// Check if circuit is supported
    pub fn supports_circuit(&self, circuit: CircuitType) -> bool {
        self.circuits.contains_key(&circuit)
    }

    /// Get all supported circuits
    pub fn supported_circuits(&self) -> Vec<CircuitType> {
        self.circuits.keys().copied().collect()
    }

    /// Get circuits by mode
    pub fn circuits_by_mode(&self, mode: ProofMode) -> Vec<CircuitType> {
        self.circuits
            .iter()
            .filter(|(_, config)| config.default_mode == mode)
            .map(|(id, _)| *id)
            .collect()
    }

    /// Get total proving key size for a set of circuits
    pub fn total_proving_key_size(&self, circuits: &[CircuitType]) -> u64 {
        circuits
            .iter()
            .filter_map(|c| self.circuits.get(c))
            .map(|c| c.proving_key_size)
            .sum()
    }

    /// Get CDN path for proving key
    pub fn proving_key_path(&self, circuit: CircuitType, cdn_base: &str) -> Option<String> {
        self.circuits.get(&circuit).map(|config| {
            let name = config.name.to_lowercase().replace(' ', "-");
            format!("{}/{}/proving-key.bin", cdn_base, name)
        })
    }
}

impl Default for CircuitRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_registry_initialization() {
        let registry = CircuitRegistry::new();

        assert!(registry.supports_circuit(CircuitType::AiInference));
        assert!(registry.supports_circuit(CircuitType::PrivacyWithdraw));
        assert!(registry.supports_circuit(CircuitType::MerkleMembership));
    }

    #[test]
    fn test_circuits_by_mode() {
        let registry = CircuitRegistry::new();

        let gpu_circuits = registry.circuits_by_mode(ProofMode::WorkerGpu);
        assert!(gpu_circuits.contains(&CircuitType::AiInference));
        assert!(!gpu_circuits.contains(&CircuitType::PrivacyWithdraw));

        let tee_circuits = registry.circuits_by_mode(ProofMode::TeeAssisted);
        assert!(tee_circuits.contains(&CircuitType::PrivacyWithdraw));
        assert!(!tee_circuits.contains(&CircuitType::AiInference));
    }
}
